[#7] 개발환경에 데이터베이스 연동 + prisma orm 설정

woobuntu opened the pull request.

woobuntu committed:
데이터베이스 컨테이너 추가

- 데이터베이스 컨테이너 -> 서버 컨테이너 -> 클라이언트 컨테이너 순의 종속성 설정
- 데이터베이스 컨테이너는 오직 서버 컨테이너에만 노출하기 위해 private network 설정

woobuntu committed:
prisma 설정

- 서버에서는 prisma orm을 사용
- 서버 도커 파일에서 prisma 관련 내용 추가
- 도커 컴포즈에서 prisma bind mount + 커넥션 url 환경변수 파일로 전달

choru-k commented on docker-compose-dev.yml:
```diff
@@ -1,20 +1,46 @@
-      - ./client/package.json:/app/package.json:rw
-      - ./client/package-lock.json:/app/package-lock.json:rw
-    ports:
-      - 3000:3000
+      - double-entry-book-keeping-data:/var/lib/mysql:rw
+    env_file:
+      - .database.env
+    networks:
+      - private
   server:
     build:
       context: ./server
       target: development
     volumes:
       - ./server/src:/app/src:rw
+      - ./server/prisma:/app/prisma:rw
       - ./server/package.json:/app/package.json:rw
       - ./server/package-lock.json:/app/package-lock.json:rw
     ports:
-      - 4000:3000
\ No newline at end of file
+      - 4000:3000
+    depends_on:
+      - database
+    networks:
+      - private
+      - default
+    env_file:
+      - ./server/.env
```
만약 health check 를 추가한다면 어떻게 할 수 있을까요?
db 가 잘 켜졌다는걸 server 에서 보장할 수 있을까요?

woobuntu replied:
@g6ling 일단 [데이터베이스 컨테이너의 경우 별도의 헬스 체크 셸 스크립트가 있어 해당 스크립트로 헬스 체크를 하게끔 수정](https://github.com/f-lab-edu/double-entry-book-keeping/pull/8/commits/09b37c8a46999e00757447ccd1905817fbe03a83)하였습니다. 그리고 [서버 컨테이너에는 헬스 체크 엔드포인트를 만들어서 해당 엔드포인트로 헬스체크를 하도록 만들었습니다](https://github.com/f-lab-edu/double-entry-book-keeping/pull/8/commits/9a297b50d600b844c109af69b93156a1ab8a1d74). 추가로 서버 컨테이너의 헬스 체크 안에서 데이터베이스 커넥션 등 디펜던시 체크를 할 수 있게끔 되어 있어(단순히 SELECT 1을 하는 것이지만), 별도의 디펜던시 체크 로직을 서버 이미지에 추가하지 않았습니다. 그리고 헬스 체크와 디펜던시 체크가 되는 만큼 서버 컨테이너에서 데이터베이스 컨테이너를 depends_on 하는 부분을 삭제하였습니다.

woobuntu committed:
Merge branch 'feature/5' of https://github.com/f-lab-edu/double-entry-book-keeping into feature/7

woobuntu committed:
데이터베이스 컨테이너 헬스 체크 추가

- https://github.com/docker-library/healthcheck/tree/master/mysql 로부터 Dockerfile과 docker-healtcheck 참조
- docker-healthcheck의 10번째 줄은 원래 host="$(hostname --ip-address || echo '127.0.0.1')"였으나 hostname 커맨드가 없어서 수정
- healtcheck interval, timeout, retries 등은 임의 설정

woobuntu committed:
서버 컨테이너 헬스 체크 추가

- nestjs의 terminus 패키지로 헬스 체크 엔드포인트 추가
- 해당 헬스 체크에서 데이터베이스가 연결되었는지, 즉 디펜던시 체크도 가능한 부분이라 서버 이미지에 달리 디펜던시 체크를 추가하지 않았다.
- 또한, 헬스 체크에 디펜던시 체크가 이미 존재하고 있어 서버 컨테이너의 depends_on 구문도 삭제하였다.
